internal/backend/triggers/triggers.go

```diff
@@ -3,45 +3,89 @@ package triggers
 import (
 	"context"
 	"errors"
+	"fmt"
 
+	"go.temporal.io/sdk/client"
 	"go.uber.org/zap"
 
 	"go.autokitteh.dev/autokitteh/internal/backend/db"
+	"go.autokitteh.dev/autokitteh/internal/backend/dispatcher"
+	"go.autokitteh.dev/autokitteh/internal/backend/temporalclient"
+	"go.autokitteh.dev/autokitteh/internal/kittehs"
 	"go.autokitteh.dev/autokitteh/sdk/sdkerrors"
 	"go.autokitteh.dev/autokitteh/sdk/sdkservices"
 	"go.autokitteh.dev/autokitteh/sdk/sdktypes"
 )
 
 type triggers struct {
-	z  *zap.Logger
-	db db.DB
+	z      *zap.Logger
+	db     db.DB
+	tmprl  temporalclient.Client
+	events sdkservices.Events
 }
 
-func New(z *zap.Logger, db db.DB) sdkservices.Triggers {
-	return &triggers{db: db, z: z}
+func New(z *zap.Logger, db db.DB, t temporalclient.Client, e sdkservices.Events) sdkservices.Triggers {
+	return &triggers{db: db, z: z, tmprl: t, events: e}
 }
 
 func (m *triggers) Create(ctx context.Context, trigger sdktypes.Trigger) (sdktypes.TriggerID, error) {
 	if trigger.ID().IsValid() {
 		return sdktypes.InvalidTriggerID, errors.New("trigger id already defined")
 	}
-
 	trigger = trigger.WithNewID()
 
-	if err := m.db.CreateTrigger(ctx, trigger); err != nil {
-		return sdktypes.InvalidTriggerID, err
+	var err error
+	if schedule, found := trigger.Data()[sdktypes.ScheduleDataSection]; found && schedule.IsValid() {
+		trigger, err = m.createScheduledWorkflow(ctx, trigger, schedule)
 	}
 
+	if err == nil {
+		err = m.db.CreateTrigger(ctx, trigger)
+	}
+
+	if err != nil {
+		return sdktypes.InvalidTriggerID, err
+	}
 	return trigger.ID(), nil
 }
 
 func (m *triggers) Update(ctx context.Context, trigger sdktypes.Trigger) error {
-	return m.db.UpdateTrigger(ctx, trigger)
+	prevTrigger, err := m.db.GetTrigger(ctx, trigger.ID())
+	if err != nil {
+		return err
+	}
+
+	prevData := prevTrigger.Data()
+	data := trigger.Data()
+	schedule, isSchedulerTrigger := data[sdktypes.ScheduleDataSection]
+	prevSchedule, isSchedulerPrevTrigger := prevData[sdktypes.ScheduleDataSection]
+
+	if !isSchedulerPrevTrigger && isSchedulerTrigger { // trigger -> scheduler trigger
+		trigger, err = m.createScheduledWorkflow(ctx, trigger, schedule)
+	} else if isSchedulerPrevTrigger && !isSchedulerTrigger { // scheduler trigger -> trigger
+		err = m.deleteSchedulerWorkflow(ctx, prevData)
+	} else if isSchedulerPrevTrigger && isSchedulerTrigger { // scheduler trigger -> scheduler trigger
+		if schedule.String() != prevSchedule.String() { // schedule changed
+			trigger, err = m.updateSchedulerWorkflow(ctx, trigger, prevData, schedule.String())
+		}
+	}
+
+	// REVIEW: should we update trigger if scheduler workflow op failed?
```

Comment:
Is this a task before merging, or does this need a ticket?

Reply:
it's an open question. What's the behavior we want?

Reply:
rewritten